File: backend/app/services/matching.py
"""The matching engine.

One reusable service, used by the student match list, the company's ranked
candidate list, and the score stored on an application. The calculation is fixed
by the brief and is meant to be explainable line by line:

    Match Score = 60% weighted skill compatibility
                + 20% CGPA eligibility
                + 10% project signal
                + 10% location / preference fit
"""

from __future__ import annotations

from dataclasses import dataclass, field

from app.models import Job, Student

SKILL_WEIGHT = 0.60
CGPA_WEIGHT = 0.20
PROJECT_WEIGHT = 0.10
LOCATION_WEIGHT = 0.10

STATUS_STRENGTH = "Strength"
STATUS_GAP = "Gap"


@dataclass
class SkillBreakdown:
    skill: str
    studentScore: int
    requiredScore: int
    gap: int
    status: str
    weight: float


@dataclass
class MatchResult:
    matchScore: int
    eligible: bool
    reasons: list[str] = field(default_factory=list)
    skillBreakdown: list[SkillBreakdown] = field(default_factory=list)
    strengths: list[str] = field(default_factory=list)
    gaps: list[str] = field(default_factory=list)
    studentCgpa: float = 0.0
    requiredCgpa: float = 0.0
    components: dict[str, float] = field(default_factory=dict)


def _skill_scores(student: Student) -> dict[str, float]:
    return {link.skill.name: link.final_score for link in student.skills}


def _skill_compatibility(
    student_scores: dict[str, float], job: Job
) -> tuple[float, list[SkillBreakdown]]:
    """Weighted attainment against every required skill, 0.0–1.0."""
    breakdown: list[SkillBreakdown] = []
    total_weight = 0.0
    attained = 0.0

    for requirement in job.required_skills:
        name = requirement.skill.name
        student_score = student_scores.get(name, 0.0)
        required = requirement.required_score

        skill_fit = min(student_score / required, 1.0) if required > 0 else 1.0
        gap = max(required - student_score, 0.0)

        attained += skill_fit * requirement.weight
        total_weight += requirement.weight

        breakdown.append(
            SkillBreakdown(
                skill=name,
                studentScore=round(student_score),
                requiredScore=round(required),
                gap=round(gap),
                status=STATUS_GAP if gap > 0 else STATUS_STRENGTH,
                weight=requirement.weight,
            )
        )

    compatibility = attained / total_weight if total_weight else 0.0
    breakdown.sort(key=lambda row: (row.status != STATUS_GAP, -row.gap))
    return compatibility, breakdown


def _cgpa_component(student: Student, job: Job) -> tuple[float, bool]:
    """Full marks when the cut-off is met, tapering to zero two points below."""
    if job.min_cgpa <= 0:
        return 1.0, True
    if student.cgpa >= job.min_cgpa:
        return 1.0, True
    shortfall = job.min_cgpa - student.cgpa
    return max(0.0, 1.0 - shortfall / 2.0), False


def _project_component(student: Student, job: Job) -> tuple[float, int, int]:
    """How much of the job's skill list the student has actually built with."""
    required = [requirement.skill.name.lower() for requirement in job.required_skills]
    if not required:
        return 0.0, 0, 0

    evidence = " ".join(
        f"{project.skills_text} {project.title}".lower() for project in student.projects
    )
    matched = sum(1 for name in required if name in evidence)

    if matched:
        return matched / len(required), matched, len(required)
    # Having built anything at all is worth a little, even off-target.
    return (0.2 if student.projects else 0.0), 0, len(required)


def _location_component(student: Student, job: Job) -> tuple[float, str]:
    preferred = (student.preferred_city or "").strip().lower()
    location = (job.location or "").strip().lower()

    if not preferred:
        return 0.5, "No city preference set"
    if location in {"remote", "anywhere"}:
        return 1.0, "Role is remote"
    if preferred == location:
        return 1.0, f"Preferred city matches {job.location}"
    return 0.0, f"Prefers {student.preferred_city}, role is in {job.location}"


def match_student_to_job(student: Student, job: Job) -> MatchResult:
    """Score one student against one job and explain the result."""
    student_scores = _skill_scores(student)

    compatibility, breakdown = _skill_compatibility(student_scores, job)
    cgpa_fit, eligible = _cgpa_component(student, job)
    project_fit, matched_projects, required_count = _project_component(student, job)
    location_fit, location_reason = _location_component(student, job)

    score = (
        compatibility * SKILL_WEIGHT
        + cgpa_fit * CGPA_WEIGHT
        + project_fit * PROJECT_WEIGHT
        + location_fit * LOCATION_WEIGHT
    ) * 100

    strengths = [row.skill for row in breakdown if row.status == STATUS_STRENGTH]
    gaps = [row.skill for row in breakdown if row.status == STATUS_GAP]

    reasons: list[str] = []

    for row in sorted(
        (row for row in breakdown if row.status == STATUS_STRENGTH),
        key=lambda row: -row.studentScore,
    )[:2]:
        reasons.append(f"Strong {row.skill} score: {row.studentScore}/100")

    if eligible:
        reasons.append(
            f"Meets CGPA requirement: {student.cgpa} / {job.min_cgpa}"
        )
    else:
        reasons.append(
            f"Below CGPA requirement: {student.cgpa} / {job.min_cgpa}"
        )

    for row in sorted(
        (row for row in breakdown if row.status == STATUS_GAP), key=lambda row: -row.gap
    )[:2]:
        reasons.append(
            f"{row.skill} gap: required {row.requiredScore}, current score {row.studentScore}"
        )

    if required_count:
        if matched_projects:
            reasons.append(
                f"Projects evidence {matched_projects} of {required_count} required skills"
            )
        elif student.projects:
            reasons.append("Has projects on record, none matching the required skills")
        else:
            reasons.append("No projects on record")

    reasons.append(location_reason)

    return MatchResult(
        matchScore=round(score),
        eligible=eligible,
        reasons=reasons,
        skillBreakdown=breakdown,
        strengths=strengths,
        gaps=gaps,
        studentCgpa=student.cgpa,
        requiredCgpa=job.min_cgpa,
        components={
            "skillCompatibility": round(compatibility * 100, 1),
            "cgpaEligibility": round(cgpa_fit * 100, 1),
            "projectSignal": round(project_fit * 100, 1),
            "locationFit": round(location_fit * 100, 1),
        },
    )


def rank_students_for_job(students: list[Student], job: Job) -> list[tuple[Student, MatchResult]]:
    """Every student scored against one job, best first."""
    scored = [(student, match_student_to_job(student, job)) for student in students]
    scored.sort(key=lambda pair: (-pair[1].matchScore, -pair[0].cgpa))
    return scored


def rank_jobs_for_student(student: Student, jobs: list[Job]) -> list[tuple[Job, MatchResult]]:
    """Every job scored for one student, best first."""
    scored = [(job, match_student_to_job(student, job)) for job in jobs]
    scored.sort(key=lambda pair: -pair[1].matchScore)
    return scored

File: frontend/src/lib/matching.ts
/**
 * SkillBridge — scoring and matching logic.
 *
 * Everything here is pure: given a student and a job it returns the same
 * result every time. That is what makes the demo honest — when the student's
 * skill scores change after the test, every match percentage recomputes from
 * the new numbers instead of flipping to a second hardcoded value.
 */

import type {
  ApplicationStatus,
  AssessedSkill,
  CareerReadiness,
  Candidate,
  GapSeverity,
  Job,
  JobMatch,
  SkillComparison,
  SkillLevel,
  SkillName,
  SkillScore,
  SkillStatus,
  Student,
} from "./types";

/* ------------------------------------------------------------------ */
/* Small helpers                                                       */
/* ------------------------------------------------------------------ */

export function clamp(value: number, min = 0, max = 100): number {
  return Math.max(min, Math.min(max, value));
}

export function levelFor(score: number): SkillLevel {
  if (score >= 80) return "Advanced";
  if (score >= 65) return "Proficient";
  if (score >= 45) return "Developing";
  return "Beginner";
}

export function severityFor(gap: number): GapSeverity {
  if (gap >= 20) return "CRITICAL";
  if (gap >= 8) return "MODERATE";
  return "HEALTHY";
}

export function statusFor(studentScore: number, requiredScore: number): SkillStatus {
  if (studentScore >= requiredScore + 10) return "STRENGTH";
  if (studentScore >= requiredScore) return "MET";
  return "GAP";
}

export function scoreOf(skills: SkillScore[], skill: SkillName): number {
  return skills.find((s) => s.skill === skill)?.score ?? 0;
}

/* ------------------------------------------------------------------ */
/* Career readiness                                                    */
/* ------------------------------------------------------------------ */

export function computeReadiness(
  skills: SkillScore[],
  cgpa: number,
  assessmentCompleted: boolean,
  profileCompletion: number,
): CareerReadiness {
  const assessed = skills.filter((s) => s.source === "Assessment");
  const pool = assessed.length > 0 ? assessed : skills;
  const avgSkill = pool.length
    ? pool.reduce((sum, s) => sum + s.score, 0) / pool.length
    : 0;

  const academic = (cgpa / 10) * 100;
  const raw =
    avgSkill * 0.6 +
    academic * 0.25 +
    profileCompletion * 0.1 +
    (assessmentCompleted ? 5 : 0);

  const score = Math.round(clamp(raw));

  let label = "Getting started";
  let summary = "Take the skill assessment to unlock accurate job matches.";
  if (score >= 80) {
    label = "Placement ready";
    summary = "Your profile clears the bar for most roles on the platform.";
  } else if (score >= 65) {
    label = "Nearly ready";
    summary = "Close two skill gaps and you clear the bar for senior openings.";
  } else if (score >= 50) {
    label = "Building up";
    summary = "Core skills are forming. Focus on your two weakest areas.";
  }

  return { score, label, summary };
}

export function profileCompletionFor(student: {
  profile: { projects: unknown[]; preferredCity: string; expectedStipend: number };
  skills: SkillScore[];
  assessmentCompleted: boolean;
}): number {
  let filled = 4; // registration, name, department, academics come from college
  const total = 8;
  if (student.profile.projects.length > 0) filled += 1;
  if (student.profile.preferredCity) filled += 1;
  if (student.profile.expectedStipend > 0) filled += 1;
  if (student.assessmentCompleted) filled += 1;
  return Math.round((filled / total) * 100);
}

/* ------------------------------------------------------------------ */
/* Job matching                                                        */
/* ------------------------------------------------------------------ */

export function compareSkills(job: Job, skills: SkillScore[]): SkillComparison[] {
  return job.requiredSkills.map((requirement) => {
    const studentScore = scoreOf(skills, requirement.skill);
    const gap = Math.max(0, requirement.requiredScore - studentScore);
    return {
      skill: requirement.skill,
      studentScore,
      requiredScore: requirement.requiredScore,
      gap,
      weight: requirement.weight,
      status: statusFor(studentScore, requirement.requiredScore),
    };
  });
}

/**
 * Weighted match, 0–100.
 *
 * Two terms, because "meets every requirement" and "clears every requirement
 * comfortably" are not the same candidate. The first term is attainment against
 * the bar and carries almost all the weight; the second gives a small credit for
 * the margin above it, which stops strong candidates from bunching at 100 and
 * makes the ranking order meaningful.
 */
const SURPLUS_RANGE = 30;
const SURPLUS_WEIGHT = 0.08;

export function matchScoreFor(comparisons: SkillComparison[], cgpaMet: boolean): number {
  if (comparisons.length === 0) return 0;

  const totalWeight = comparisons.reduce((sum, c) => sum + c.weight, 0);

  const attained = comparisons.reduce((sum, c) => {
    const ratio = c.requiredScore === 0 ? 1 : c.studentScore / c.requiredScore;
    return sum + Math.min(1, ratio) * c.weight;
  }, 0);

  const surplus = comparisons.reduce((sum, c) => {
    const margin = Math.max(0, c.studentScore - c.requiredScore);
    return sum + Math.min(1, margin / SURPLUS_RANGE) * c.weight;
  }, 0);

  const base = (attained / totalWeight) * 100;
  const bonus = (surplus / totalWeight) * 100;
  const score = base * (1 - SURPLUS_WEIGHT) + bonus * SURPLUS_WEIGHT;

  return Math.round(clamp(cgpaMet ? score : score - 12));
}

function buildReason(comparisons: SkillComparison[], cgpaMet: boolean): string {
  const strong = comparisons
    .filter((c) => c.status !== "GAP")
    .sort((a, b) => b.studentScore - b.requiredScore - (a.studentScore - a.requiredScore))
    .slice(0, 2)
    .map((c) => c.skill);

  const worst = comparisons
    .filter((c) => c.status === "GAP")
    .sort((a, b) => b.gap * b.weight - a.gap * a.weight)[0];

  const parts: string[] = [];
  if (strong.length === 2) {
    parts.push(`${strong[0]} and ${strong[1]} clear the bar comfortably`);
  } else if (strong.length === 1) {
    parts.push(`${strong[0]} clears the bar`);
  } else {
    parts.push("No required skill has reached the bar yet");
  }

  if (worst) {
    parts.push(`${worst.skill} is ${worst.gap} points short`);
  } else {
    parts.push("every required skill is met");
  }

  if (!cgpaMet) parts.push("CGPA is below the posted minimum");

  return `${parts.join(", ")}.`;
}

export function buildJobMatch(
  job: Job,
  student: Student,
  applied: boolean,
  applicationId: string | null,
): JobMatch {
  const comparisons = compareSkills(job, student.skills);
  const cgpaMet = student.academics.cgpa >= job.minCgpa;
  const matchScore = matchScoreFor(comparisons, cgpaMet);

  return {
    job,
    matchScore,
    strengths: comparisons.filter((c) => c.status !== "GAP").map((c) => c.skill),
    gaps: comparisons.filter((c) => c.status === "GAP").map((c) => c.skill),
    reason: buildReason(comparisons, cgpaMet),
    comparisons,
    cgpaMet,
    applied,
    applicationId,
  };
}

/* ------------------------------------------------------------------ */
/* Candidate ranking (company side)                                    */
/* ------------------------------------------------------------------ */

function candidateReasons(
  comparisons: SkillComparison[],
  student: Student,
  job: Job,
): string[] {
  const reasons: string[] = [];

  const strengths = comparisons.filter((c) => c.status === "STRENGTH");
  if (strengths.length > 0) {
    reasons.push(
      `Exceeds the bar on ${strengths
        .map((c) => `${c.skill} (${c.studentScore} vs ${c.requiredScore})`)
        .join(", ")}.`,
    );
  }

  const met = comparisons.filter((c) => c.status === "MET");
  if (met.length > 0) {
    reasons.push(`Meets the requirement on ${met.map((c) => c.skill).join(", ")}.`);
  }

  const gaps = comparisons
    .filter((c) => c.status === "GAP")
    .sort((a, b) => b.gap - a.gap);
  if (gaps.length > 0) {
    reasons.push(
      `Needs development in ${gaps
        .map((c) => `${c.skill} (${c.gap} points short)`)
        .join(", ")}.`,
    );
  }

  reasons.push(
    student.academics.cgpa >= job.minCgpa
      ? `CGPA ${student.academics.cgpa.toFixed(1)} clears the ${job.minCgpa.toFixed(1)} cut-off.`
      : `CGPA ${student.academics.cgpa.toFixed(1)} is below the ${job.minCgpa.toFixed(1)} cut-off.`,
  );

  if (student.profile.projects.length > 0) {
    reasons.push(
      `${student.profile.projects.length} verified project${
        student.profile.projects.length > 1 ? "s" : ""
      } on record, most recently "${student.profile.projects[0].title}".`,
    );
  }

  return reasons;
}

export function rankCandidates(
  job: Job,
  students: Student[],
  lookupApplication: (
    studentId: string,
  ) => { status: ApplicationStatus; applicationId: string } | null,
): Candidate[] {
  return students
    .map((student) => {
      const comparisons = compareSkills(job, student.skills);
      const cgpaMet = student.academics.cgpa >= job.minCgpa;
      const application = lookupApplication(student.id);

      return {
        studentId: student.id,
        name: student.name,
        department: student.academics.department,
        semester: student.academics.semester,
        cgpa: student.academics.cgpa,
        college: student.academics.college,
        matchScore: matchScoreFor(comparisons, cgpaMet),
        rank: 0,
        strengths: comparisons.filter((c) => c.status !== "GAP").map((c) => c.skill),
        gaps: comparisons.filter((c) => c.status === "GAP").map((c) => c.skill),
        comparisons,
        skills: student.skills,
        projects: student.profile.projects,
        reasons: candidateReasons(comparisons, student, job),
        status: application?.status ?? null,
        applicationId: application?.applicationId ?? null,
        applied: application !== null,
      };
    })
    .sort((a, b) => b.matchScore - a.matchScore || b.cgpa - a.cgpa)
    .map((candidate, index) => ({ ...candidate, rank: index + 1 }));
}

/* ------------------------------------------------------------------ */
/* Assessment scoring                                                  */
/* ------------------------------------------------------------------ */

/**
 * Applies an assessment result to an existing skill score.
 *
 * A two-question sample is far too small to replace a score outright, so the
 * result moves the existing score by a delta instead. Lower scores move more
 * than high ones (there is more headroom), which keeps the update intuitive:
 * full marks always improves a skill, a miss always costs a little.
 */
export function updatedSkillScore(
  previousScore: number,
  correct: number,
  total: number,
): number {
  const ratio = total === 0 ? 0 : correct / total;

  let delta: number;
  if (ratio >= 1) {
    delta = Math.round(4 + (100 - previousScore) * 0.04);
  } else if (ratio > 0) {
    delta = Math.round(ratio * (2 + (100 - previousScore) * 0.02));
  } else {
    delta = -Math.round(3 + previousScore * 0.03);
  }

  return Math.round(clamp(previousScore + delta, 0, 98));
}

export function strongestAndWeakest(
  scores: { skill: AssessedSkill; score: number }[],
): { strongest: AssessedSkill[]; weakest: AssessedSkill[] } {
  const sorted = [...scores].sort((a, b) => b.score - a.score);
  return {
    strongest: sorted.slice(0, 2).map((s) => s.skill),
    weakest: sorted.slice(-2).reverse().map((s) => s.skill),
  };
}
